File: cpp/tools/parquet/parquet_reader.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include <iostream>
#include <list>
#include <memory>

#include "parquet/api/reader.h"

int main(int argc, char** argv) {
  if (argc > 5 || argc < 2) {
    std::cerr << "Usage: parquet_reader [--only-metadata] [--no-memory-map] "
                 "[--columns=...] <file>"
              << std::endl;
    return -1;
  }

  std::string filename;
  bool print_values = true;
  bool memory_map = true;

  // Read command-line options
  const std::string COLUMNS_PREFIX = "--columns=";
  std::list<int> columns;

  char *param, *value;
  for (int i = 1; i < argc; i++) {
    if ((param = std::strstr(argv[i], "--only-metadata"))) {
      print_values = false;
    } else if ((param = std::strstr(argv[i], "--no-memory-map"))) {
      memory_map = false;
    } else if ((param = std::strstr(argv[i], COLUMNS_PREFIX.c_str()))) {
      value = std::strtok(param + COLUMNS_PREFIX.length(), ",");
      while (value) {
        columns.push_back(std::atoi(value));
        value = std::strtok(nullptr, ",");
      }
    } else {
      filename = argv[i];
    }
  }

  try {
    std::unique_ptr<parquet::ParquetFileReader> reader =
        parquet::ParquetFileReader::OpenFile(filename, memory_map);
    parquet::ParquetFilePrinter printer(reader.get());
    printer.DebugPrint(std::cout, columns, print_values);
  } catch (const std::exception& e) {
    std::cerr << "Parquet error: " << e.what() << std::endl;
    return -1;
  }

  return 0;
}

File: java/adapter/parquet/src/main/java/org/apache/arrow/adapter/parquet/ParquetReader.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.arrow.adapter.parquet;

import java.io.IOException;
import java.util.List;

import org.apache.arrow.adapter.common.ArrowRecordBatchBuilder;
import org.apache.arrow.adapter.common.ArrowRecordBatchBuilderImpl;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.VectorLoader;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.ipc.ReadChannel;
import org.apache.arrow.vector.ipc.message.ArrowRecordBatch;
import org.apache.arrow.vector.ipc.message.MessageChannelReader;
import org.apache.arrow.vector.ipc.message.MessageResult;
import org.apache.arrow.vector.ipc.message.MessageSerializer;
import org.apache.arrow.vector.types.pojo.Schema;
import org.apache.arrow.vector.util.ByteArrayReadableSeekableByteChannel;

/** Parquet Reader Class. */
public class ParquetReader implements AutoCloseable {

  /** reference to native reader instance. */
  private long nativeInstanceId;

  /** last readed length of a record batch. */
  private long lastReadLength;

  private BufferAllocator allocator;
  private ParquetReaderJniWrapper jniWrapper;

  /**
   * Create an instance for ParquetReader.
   *
   * @param path Parquet Reader File Path.
   * @param rowGroupIndices An array to indicate which rowGroup to read.
   * @param columnIndices An array to indicate which columns to read.
   * @param batchSize number of rows expected to be read in one batch.
   * @param allocator A BufferAllocator reference.
   * @throws IOException throws io exception in case of native failure.
   */
  public ParquetReader(
      String path,
      int[] rowGroupIndices,
      int[] columnIndices,
      long batchSize,
      BufferAllocator allocator)
      throws IOException {
    this.jniWrapper = new ParquetReaderJniWrapper();
    this.allocator = allocator;
    this.nativeInstanceId = jniWrapper.nativeOpenParquetReader(path, batchSize);
    jniWrapper.nativeInitParquetReader(nativeInstanceId, columnIndices, rowGroupIndices);
  }

  /**
   * Create an instance for ParquetReader.
   *
   * @param path Parquet Reader File Path.
   * @param startPos A start pos to indicate which rowGroup to read.
   * @param endPos An end pos indicate which rowGroup to read.
   * @param columnIndices An array to indicate which columns to read.
   * @param batchSize number of rows expected to be read in one batch.
   * @param allocator A BufferAllocator reference.
   * @throws IOException throws io exception in case of native failure.
   */
  public ParquetReader(
      String path,
      long startPos,
      long endPos,
      int[] columnIndices,
      long batchSize,
      BufferAllocator allocator)
      throws IOException {
    this.jniWrapper = new ParquetReaderJniWrapper();
    this.allocator = allocator;
    this.nativeInstanceId = jniWrapper.nativeOpenParquetReader(path, batchSize);
    jniWrapper.nativeInitParquetReader2(nativeInstanceId, columnIndices, startPos, endPos);
  }

  /**
   * Get Arrow Schema from ParquetReader.
   *
   * @return Schema of parquet file
   * @throws IOException throws io exception in case of native failure
   */
  Schema getSchema() throws IOException {
    byte[] schemaBytes = jniWrapper.nativeGetSchema(nativeInstanceId);

    try (MessageChannelReader schemaReader =
        new MessageChannelReader(
            new ReadChannel(new ByteArrayReadableSeekableByteChannel(schemaBytes)), allocator)) {

      MessageResult result = schemaReader.readNext();
      if (result == null) {
        throw new IOException("Unexpected end of input. Missing schema.");
      }

      return MessageSerializer.deserializeSchema(result.getMessage());
    }
  }

  /**
   * Read Next ArrowRecordBatch from ParquetReader.
   *
   * @return One ArrowRecordBatch readed from parquet file reader
   * @throws IOException throws io exception in case of native failure
   */
  ArrowRecordBatch readNext() throws IOException {
    ArrowRecordBatchBuilder recordBatchBuilder = jniWrapper.nativeReadNext(nativeInstanceId);
    if (recordBatchBuilder == null) {
      return null;
    }
    ArrowRecordBatchBuilderImpl recordBatchBuilderImpl =
        new ArrowRecordBatchBuilderImpl(recordBatchBuilder);
    ArrowRecordBatch batch = recordBatchBuilderImpl.build();
    this.lastReadLength = batch.getLength();
    return batch;
  }

  /**
   * Read Next ValueVectorList from ParquetReader.
   *
   * @return Next ValueVectorList readed from parquet file.
   * @throws IOException throws io exception in case of native failure
   */
  public List<FieldVector> readNextVectors(VectorSchemaRoot root) throws IOException {
    ArrowRecordBatch batch = readNext();
    if (batch == null) {
      return null;
    }
    VectorLoader loader = new VectorLoader(root);
    loader.load(batch);
    batch.close();
    return root.getFieldVectors();
  }

  /**
   * Get last readed ArrowRecordBatch Length.
   *
   * @return lastReadLength.
   */
  public long lastReadLength() {
    return lastReadLength;
  }

  @Override
  public void close() {
    jniWrapper.nativeCloseParquetReader(nativeInstanceId);
  }
}
